### PR TITLE
Add epoch to parameters, add methods to transports to provide epoch, sharedrandom

### PR DESCRIPTION
**Pull request by mixmasala**
this PR facilitates publishing the epoch parameters by adding an epoch parameter that will be included in the consensus (see: https://github.com/katzenpost/server/pull/144, which is needed in order to rotate the epochs each descriptor publication)

### CODE REVIEW

**Comment by mixmasala** [comment]
Note: I think this can be improved substantially by switching to using a different clock than the katzenpost clock and using longer ([]byte vs uint64) epoch values.

**Comment by mixmasala** [comment]
I realized that session resumption requires more logic; the commits above take a stab at it.

File: client/exchange.go [hunk 1]
```diff
@@ -125,7 +125,43 @@ func NewExchangeFromSnapshot(
...
+	if !current {
+		return ex, errors.New("Epoch has expired, cannot resume exchange")
+	}
+	current = false
+
+	// Verify that the session shared random is still valid
+	ssrv := ex.session.SharedRandom()
+	for _, srv := range srvs {
+		if bytes.Equal(srv, ssrv) {
+			current = true
```
Comment (david415) [inline client/exchange.go:158]: i think we can add a break after current = true to exit for loop early, no?

**Comment by david415** [comment]
> Note: I think this can be improved substantially by switching to using a different clock than the katzenpost clock and using longer ([]byte vs uint64) epoch values.
> 
> edit: the PR also adds an event that is sent if the exchange fails.

instead of switching clocks away from the katzenpost clock... we should just fix the current reunion-katzenpost-clock... i don't want a bunch of useless clocks laying around.

File: client/exchange.go [hunk 2]
```diff
@@ -510,19 +510,29 @@ func (e *Exchange) processT3Messages() bool {
...
+			Error:     errors.New("Run was halted."),
+		}
+	}
+
 	switch e.status {
 	case initialState:
 		// XXX not required -> 1:A <- DB: fetch current epoch and current set of data for epoch state
 		// 2:A -> DB: transmit א message
 		if !e.sendT1() {
+			defer haltedfn()
```
Comment (david415) [inline client/exchange.go]: why defer haledfn?

**Pull request merged**